### PR TITLE
Universal Package task should store version number in env variable

### PR DESCRIPTION
**Pull request by halfpeaw**
Universal Packages can be uploaded without a predetermined version number, so save this information in an environment variable prior to completion

### CODE REVIEW

File: Tasks/UniversalPackagesV0/universalpublish.ts [hunk 1]
```diff
@@ -153,6 +153,8 @@ function publishPackageUsingArtifactTool(publishDir: string, options: artifactTo
     const execResult: IExecSyncResult = artifactToolRunner.runArtifactTool(options.artifactToolPath, command, execOptions);
 
     if (execResult.code === 0) {
+        console.log(`##vso[task.setvariable variable=UniversalPackage_Name;]${options.packageName}`);
```
Comment (zjrunner) [inline Tasks/UniversalPackagesV0/universalpublish.ts]: I've always had a concern about the singleton nature of the task, and wanted to get a 'lock file' type of manifest supported too.  Under multi-run or multi-package, a single variable starts to lose value.  I know this is publish not download, but am still thinking a bigger "what I've done" parseable/mergable summary would be nice, but could be overkill for right now.  I don't want to get locked into permanent expectation of a singleton variable without a bit more thought though.  + @zarenner / @mitchdenny.

Comment (zarenner) [reply Tasks/UniversalPackagesV0/universalpublish.ts]: Agree - that said, there's nothing really preventing the envvar (or multiple envvars) from containing a list of packages published. Perhaps differently named. I'll let Mitch comment since I'm sure he has relevant thoughts / future plans.

Comment (mitchdenny) [reply Tasks/UniversalPackagesV0/universalpublish.ts]: To be honest, I think that these more advanced multi-upload scenarios may end up being done with scripts. You can only get so far with a set-orientated upload of multiple packages before you start needing to think about branching logic etc which all screens scripting.

At which point Universal Packages emitting JSON like the AZ CLI seems appropriate (especially if we support ```--query``` syntax.

File: Tasks/UniversalPackagesV0/task.json [hunk 2]
```diff
@@ -316,6 +316,15 @@
             "required": "false",
             "groupName": "packagePublishDetails"
         },
+        {
+            "name": "publishedPackageVar",
+            "type": "string",
+            "label": "Package Ouput Variable",
+            "required": false,
+            "defaultValue": "",
+            "groupName": "packagePublishDetails",
```
Comment (zjrunner) [inline Tasks/UniversalPackagesV0/task.json]: @mitchdenny - in the publish details group, not an output group like your other example?

File: Tasks/UniversalPackagesV0/universalpublish.ts [hunk 3]
```diff
@@ -121,6 +122,7 @@ export async function run(artifactToolPath: string): Promise<void> {
             accountUrl: serviceUri,
             packageName,
             packageVersion: version,
+            publishedPackageVar,
```
Comment (zjrunner) [inline Tasks/UniversalPackagesV0/universalpublish.ts]: Instead of passing this in, then conditionally checking for success, can you just setResult around line 130 where it is already tl.setResult(tl.Taskresult.Succeeded...)?  would cut down on changes and unify the success handling.

File: Tasks/UniversalPackagesV0/universalpublish.ts [hunk 4]
```diff
@@ -153,6 +155,9 @@ function publishPackageUsingArtifactTool(publishDir: string, options: artifactTo
     const execResult: IExecSyncResult = artifactToolRunner.runArtifactTool(options.artifactToolPath, command, execOptions);
 
     if (execResult.code === 0) {
+        if(options.publishedPackageVar) {
+            tl.setVariable(options.publishedPackageVar, `${options.packageName}.${options.packageVersion}`);
```
Comment (zjrunner) [inline Tasks/UniversalPackagesV0/universalpublish.ts]: This could be hard to parse back out - '.' is allowed in both name and version isn't it?  How about a space as i can't think of a proto where that is a valid character of the name or version?  @mitchdenny ?

Comment (halfpeaw) [reply Tasks/UniversalPackagesV0/universalpublish.ts]: I figured returning the exact name would probably be the correct thing but we could have a comma instead it would be much easier to parse.  There is a fancy regex you can use instead if we stick with the period: Regex.Split(PackageName, @"\.(?=\d)");

File: Tasks/UniversalPackagesV0/universalpublish.ts (hunk 1 above)
Comment (halfpeaw) [reply Tasks/UniversalPackagesV0/universalpublish.ts]: I've updated the code to take an input variable as part of the task to store the outputted published packaged name.  I think if we want to revisit this in the future to use a json file that seem reasonable.

File: Tasks/UniversalPackagesV0/universalpublish.ts (hunk 4 above)
Comment (halfpeaw) [reply Tasks/UniversalPackagesV0/universalpublish.ts]: Added space

File: Tasks/UniversalPackagesV0/universalpublish.ts (hunk 3 above)
Comment (halfpeaw) [reply Tasks/UniversalPackagesV0/universalpublish.ts]: Moved the code

File: Tasks/UniversalPackagesV0/Strings/resources.resjson/en-US/resources.resjson [hunk 5]
```diff
@@ -43,6 +43,8 @@
   "loc.input.help.versionPublish": "Select the custom package version.",
   "loc.input.label.packagePublishDescription": "Description",
   "loc.input.help.packagePublishDescription": "Description of the contents of this package and/or the changes made in this version of the package.",
+  "loc.input.label.publishedPackageVar": "Package Ouput Variable",
+  "loc.input.help.publishedPackageVar": "Specify a name for the output variable that is generated from the published package name and version. The variable can be consumed in subsequent tasks.",
```
[the remark this answers is not in the document]
Comment (halfpeaw) [reply Tasks/UniversalPackagesV0/Strings/resources.resjson/en-US/resources.resjson]: Fixed

File: Tasks/UniversalPackagesV0/task.json (hunk 2 above)
Comment (halfpeaw) [reply Tasks/UniversalPackagesV0/task.json]: Updated to include an output group

**Comment by mitchdenny** [comment]
@zjrunner this change looks good to me - did you want to merge it or shall I?
